clients/cfrestclient/rest_cloud_foundry_client_extended.go

```diff
@@ -1,65 +1,154 @@
 package cfrestclient
 
 import (
+	"code.cloudfoundry.org/cli/plugin"
+	"code.cloudfoundry.org/jsonry"
 	"context"
+	"crypto/md5"
+	"encoding/hex"
+	"encoding/json"
 	"fmt"
+	"io"
 	"net/http"
 	"net/url"
+	"strings"
 
-	baseclient "github.com/cloudfoundry-incubator/multiapps-cli-plugin/clients/baseclient"
+	"github.com/cloudfoundry-incubator/multiapps-cli-plugin/clients/baseclient"
 	"github.com/cloudfoundry-incubator/multiapps-cli-plugin/clients/cfrestclient/operations"
-	models "github.com/cloudfoundry-incubator/multiapps-cli-plugin/clients/models"
-	strfmt "github.com/go-openapi/strfmt"
+	"github.com/cloudfoundry-incubator/multiapps-cli-plugin/clients/models"
+	"github.com/go-openapi/strfmt"
 )
 
-const cfBaseUrl = "v2/"
-const defaultSharedDomainsPathPattern = "/shared_domains"
+const cfBaseUrl = "v3/"
+const defaultDomainsPathPattern = "/domains"
 
 type CloudFoundryRestClient struct {
 	baseclient.BaseClient
 	httpCloudFoundryClient *CloudFoundryClientExtended
+	cliConn                plugin.CliConnection
 }
 
-func NewCloudFoundryRestClient(host string, rt http.RoundTripper, tokenFactory baseclient.TokenFactory) CloudFoundryOperationsExtended {
-	t := baseclient.NewHTTPTransport(host, cfBaseUrl, rt)
+func NewCloudFoundryRestClient(cliConn plugin.CliConnection, rt http.RoundTripper, tokenFactory baseclient.TokenFactory) CloudFoundryOperationsExtended {
+	t := baseclient.NewHTTPTransport(getApiEndpoint(cliConn), cfBaseUrl, rt)
 	httpCloudFoundryClient := New(t, strfmt.Default)
-	return &CloudFoundryRestClient{baseclient.BaseClient{TokenFactory: tokenFactory}, httpCloudFoundryClient}
+	return &CloudFoundryRestClient{baseclient.BaseClient{TokenFactory: tokenFactory}, httpCloudFoundryClient, cliConn}
 }
 
-func (c CloudFoundryRestClient) GetSharedDomains() ([]models.SharedDomain, error) {
-	result := []models.SharedDomain{}
+func getApiEndpoint(cliConnection plugin.CliConnection) string {
+	api, err := cliConnection.ApiEndpoint()
+	if err != nil {
```

Comment:
In my opinion we should at least log this error

Reply:
This method will never throw an error (https://github.com/cloudfoundry/cli/blob/master/plugin/rpc/cli_rpc_server.go#L266). I just copied this method from `BaseCommand`